### ISSUE TITLE
[intro-neural-networks/student-admissions/StudentAdmissions] issues

### ISSUE THREAD

**Issue by kuzand** [udacity/deep-learning-v2-pytorch#99]
What is implemented there is a single-layer NN (input layer + one output node) and this should be stated clearly (i.e. change the "Training the 2-layer Neural Network"). 

Couple of other things:

1. The function implementing the cross-entropy loss `error_formula` is actually never used, although it appears in the `train_nn` function, line `error = error_formula(y, output)` but the `error` is unused.

2. There is no bias used in the training, only weights.
 
3. The error function used is the squared error function, `loss = np.mean((out - targets) ** 2)` and the error_term is computed as `(y - output)*sigmoid_prime(x)`, not the cross-entropy error as one would expect (see the previous lessons and the `error_formula`)
 
4. In the `plot_points` change the `s` in list comprehensions to some other variable name because it can be confusing for some students since `s = 25` is the marker size:
`plt.scatter([s[0][0] for s in rejected], [s[0][1] for s in rejected], s = 25, color = 'red', edgecolor = 'k')`

And a general question: what's the point in using single-layer perceptron for a highly non-classical problem, after introducing the multilayer neural networks and backpropagation? I think it would be better to implement the code for a NN with 1 hidden layer to demonstrate in practice what was learned in previous lesson.

**Comment by shyamsundar-naik** [comment]
Why is this issue under autoencoder section?

**Comment by kuzand** [comment]
> 
> 
> Why is this issue under autoencoder section?

I think you are confusing something.

**Comment by kuzand** [comment]
Also, the formula `(y - output)*sigmoid_prime(x)` is wrong and should be `(y - output)*output*(1 - output)`. Here's the derivation (I use y_hat = output):

Ε = 0.5 * (y - y_hat)**2
y_hat  = sigma(z)
z = W*x + b
dE/dw = (dE/dy_hat)*(dy_hat/dz)*(dz/dw) = -(y - y_hat)*y_hat*(1 - y_hat)* x
where dy_hat/dz = sigma_prime(z), not sigmoid_prime(x).

**Comment by shyamsundar-naik** [comment]
I see that this is the wrong area I have posted. I am in search of people
who work on data conversion.

**Comment by lleiou** [comment]
Finally understand why I'm unable to derive the `error term` in the answer, the original question was too misleading and messed up...

Thanks for your clarification!

**Comment by arohanajit** [comment]
What I think is `error` variable is a sort of benchmark for how the model performs. It is accounted for in backpropagation but it's not directly needed as we are already applying the backpropagation formula in `error_term_formula()` function. For those who aren't clear on how it came, please refer to 3blue1brown youtube lectures on how math of backpropagation is calculated. and the `error_formula()` is just used to calculate total error which is essentially error of final node. Link: https://www.youtube.com/playlist?list=PLZHQObOWTQDNU6R1_67000Dx_ZCJB-3pi

**Comment by tdeck** [comment]
I've run into the same confusion described here and took a long time to work through it. In addition I found these related issues when searching: https://github.com/udacity/deep-learning-v2-pytorch/issues/208 https://github.com/udacity/deep-learning-v2-pytorch/issues/75 I hope it's helpful to someone to write about my solution here - and perhaps I can get a critique if I've made a mistake.

I tried to make this work with log-loss, since that was what had been introduced in the course. Here's what I ended up with; I'm not sure it's correct but it seems to work:

1. I computed the derivative of the loss with respect to W and got what lleiou wrote above. Another way to write it is `(yhat - y) * x`. [This post](https://medium.com/@opetundeadepoju/a-step-by-step-tutorial-on-coding-neural-network-logistic-regression-model-from-scratch-5f9025bd3d6) has the same answer.
2. The code doesn't actually make it clear what an `error_term` is, you have to look later in the code to figure it out. But in this case `error_term` is dE/dw divided by X. So I used `yhat - y`
3.  In the training code, I had to replace this line `weights += learnrate * del_w / n_records` with `weights -= learnrate * del_w / n_records`, since I believe we want to go *against* the gradient of the loss (so we get *less* loss).
4. This line `loss = np.mean((out - targets) ** 2)` is using mean squared loss, a function that hasn't been introduced at all. I replaced it with `loss = np.mean(error_formula(targets, all_outputs))` (log loss) (note `all_outputs` is my new name for `out`).

Finally, there are some needlessly confusing things in the code:
- The `error` from line `error = error_formula(y, output)` isn't used at all
- I personally think `out` is a bad variable name in `out = sigmoid(np.dot(features, weights))`, because it's easily confused with the `output` from above which is still in scope. At first I accidentally passed in `output` and got weird results, as you'd expect.. So I renamed it to `all_outputs`.

I think it's really worth fixing this lesson up because I struggled with it for several hours. It's one thing to just plug in the given formulas and run the script, but if you're really trying to learn and understand all these bugs and the new loss function really get in the way.

**Comment by jaehoonLee** [comment]
+1

**Comment by abhiojha8** [comment]
This is resolved by PR #250
